src/functions.cpp

```diff
@@ -0,0 +1,169 @@
+#include "functions.h"
+
+#include "log.h"
+
+namespace bpftrace {
+
+namespace {
+std::string arg_types_str(const std::vector<SizedType> &arg_types)
+{
+  std::string str = "(";
+  bool first = true;
+  for (const SizedType &arg_type : arg_types) {
+    if (!first)
+      str += ", ";
+    str += typestr(arg_type);
+    first = false;
+  }
+  str += ")";
+  return str;
+}
+
+std::string param_types_str(const std::vector<Param> &params)
+{
+  std::string str = "(";
+  bool first = true;
+  for (const Param &param : params) {
+    if (!first)
+      str += ", ";
+    str += typestr(param.type());
+    first = false;
+  }
+  str += ")";
+  return str;
+}
+} // namespace
+
+FunctionRegistry::FunctionRegistry(std::ostream &out) : out_(out)
+{
+}
+
+const Function &FunctionRegistry::add(Function::Origin origin,
+                                      std::string_view name,
+                                      const SizedType &return_type,
+                                      const std::vector<Param> &params)
+{
+  return add(origin, {}, name, return_type, params);
+}
+
+const Function &FunctionRegistry::add(Function::Origin origin,
+                                      std::string_view ns,
+                                      std::string_view name,
+                                      const SizedType &return_type,
+                                      const std::vector<Param> &params)
+{
+  all_funcs_.push_back(std::make_unique<Function>(
+      origin, std::string{ name }, return_type, params));
+  Function &new_func = *all_funcs_.back().get();
+
+  FqName fq_name{
+    .ns = std::string{ ns },
+    .name = std::string{ name },
+  };
+
+  // Check for duplicate function definitions
+  // The assumption is that builtin functions are all added to the registry
+  // before any user-defined functions.
+  // Builtin functions can be duplicated. Other functions can not.
+  for (const Function &func : funcs_by_fq_name_[fq_name]) {
+    if (func.origin() != Function::Origin::Builtin) {
+      throw std::runtime_error("Cannot redefine existing function: " +
+                               fq_name.str() + param_types_str(params));
+    }
+  }
+
+  funcs_by_fq_name_[fq_name].push_back(new_func);
+  return new_func;
+}
+
+namespace {
+bool can_implicit_cast(const SizedType &from, const SizedType &to)
+{
+  if (from.FitsInto(to))
+    return true;
+
+  if (from.IsStringTy() && to.IsPtrTy() && to.GetPointeeTy()->IsIntTy() &&
+      to.GetPointeeTy()->GetSize() == 1) {
+    // Allow casting from string to int8* or uint8*
+    return true;
+  }
+
+  // Builtin and script functions do not care about string sizes. External
+  // functions cannot be defined to accept string types (they'd take char*)
+  if (from.IsStringTy() && to.IsStringTy())
+    return true;
+
+  return false;
+}
+} // namespace
+
+/*
+ * Find the best function called 'name' for the given argument types.
```

Comment:
```suggestion
 * Find the best function by name for the given argument types.
```
The function's name is not 'name'